### ISSUE TITLE
Support for tfrecord inputs for transformer models (and practices of large scale training with opennmt-tf)

### ISSUE THREAD

**Issue by Morizeyao** [OpenNMT/OpenNMT-tf#582]
Currently only the `opennmt.inputters.WordEmbedder` class is supported for transformer model.

I am training a quite large dataset (about 50GB raw text) with all inputs in a single file (and targets in another single file). The training is very slow in horovod distributed mode.

What I have right now are about 10 * 4 card P100 machines.

I suspect the the slowness is because of the use of single huge text files, so I want to try to implement tfrecord support for transformer models.

Is there any suggestions about this?

Thank you very much!

**Comment by guillaumekln** [comment]
How would the TFRecords accelerate the training?

What is the baseline speed on a single machine (in terms of tokens per second as reported in the training logs)?

**Comment by Morizeyao** [comment]
Thank you for the quick response!
TFRecord acceleration is my guess for now.

Baseline speed on single P100 card is 3.6 step per second.
Speed on 8 * 4 = 32 P100 cards with 4 gradient accumulation is about 0.16 step per second.
 
The speed up is 32 * 4 * 0.16 / 3.6 = 5.6 times.

However after training for about 50 steps the training progress will stuck for like 4 hours before it can train for another 50 steps. Log is like this:

```
[2020-01-08 03:13:01,788] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:248] loss = 252.08331, step = 50 (44.086 sec)
[2020-01-08 07:33:28,284] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:708] global_step/sec: 0.000639925
[2020-01-08 07:33:29,020] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:248] loss = 360.40735, step = 60 (15627.232 sec)
[2020-01-08 07:34:13,385] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:708] global_step/sec: 0.221722
[2020-01-08 07:34:13,710] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:248] loss = 272.15897, step = 70 (44.690 sec)
[2020-01-08 07:34:57,785] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:708] global_step/sec: 0.225224
[2020-01-08 07:34:58,206] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:248] loss = 251.09262, step = 80 (44.496 sec)
[2020-01-08 07:35:42,171] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:708] global_step/sec: 0.225298
[2020-01-08 07:35:42,496] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:248] loss = 205.18915, step = 90 (44.291 sec)
[2020-01-08 07:36:26,447] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:708] global_step/sec: 0.225858
[2020-01-08 07:36:26,755] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:248] loss = 132.55606, step = 100 (44.259 sec)
[2020-01-08 07:37:11,235] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:708] global_step/sec: 0.223274
[2020-01-08 07:37:11,610] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:248] loss = 171.07977, step = 110 (44.855 sec)
[2020-01-08 11:48:09,795] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:708] global_step/sec: 0.000664074
[2020-01-08 11:48:10,337] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:248] loss = 420.03006, step = 120 (15058.727 sec)
[2020-01-08 11:48:54,318] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:708] global_step/sec: 0.224601
[2020-01-08 11:48:54,709] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:248] loss = 199.54288, step = 130 (44.371 sec)
[2020-01-08 11:49:38,777] [INFO] [17#MainThread] [tensorflow/python/training/basic_session_run_hooks.py:708] global_step/sec: 0.22493
```

**Comment by guillaumekln** [comment]
Can you post the command line and configuration files?

**Comment by Morizeyao** [comment]
Sure! Opennmt-tf version is 1.25.3.

Config file is 

```
model_dir: hdfs://10.87.52.194/user/datamining/duzeyao/models/model_3_mixed_32card
iter_size: 4
log_steps: 10
data:
  train_features_file: hdfs://10.87.52.194/user/datamining/duzeyao/data/mixed/train_inputs_all.txt
  train_labels_file: hdfs://10.87.52.194/user/datamining/duzeyao/data/mixed/train_targets_all.txt
  eval_features_file: hdfs://10.87.52.194/user/datamining/duzeyao/data/mixed/train_inputs_all.txt
  eval_labels_file: hdfs://10.87.52.194/user/datamining/duzeyao/data/mixed/train_targets_all.txt


  source_words_vocabulary: hdfs://10.87.52.194/user/datamining/duzeyao/onmt_vocab/onmt_vocab_new.txt
  target_words_vocabulary: hdfs://10.87.52.194/user/datamining/duzeyao/onmt_vocab/onmt_vocab_new.txt


params:
  optimizer: AdamOptimizer
  learning_rate: 0.8

  dropout: 0.1
  decay_type: noam_decay_v2
  decay_params:
    model_dim: 1024
    warmup_steps: 5000
  start_decay_steps: 0

  minimum_learning_rate: 0
  beam_width: 5
  num_hypotheses: 5
train:
  sample_buffer_size: 125000
  batch_size: 32
  batch_type: examples
  save_checkpoints_steps: 1500
  keep_checkpoint_max: 1

  save_summary_steps: 100
  train_steps: 1000000

score:
  batch_size: 64
  with_token_level: false
  with_alignments: null
```

For the command line it is a bit complicated because the opennmt-tf code is slightly modified for the distributed system.

```
name=opennmt-1
cpu=10
memory=60000
worker_num=4
image="pai-tensorflow-emr_0.7:latest"
command="python onmt-main.py --model model.py --config 3_train_config.yaml --run=pai_distributed_train"

python bin.py --type yarn --name ${name} --cpu ${cpu} --memory ${memory} --gpus 4 --worker_num ${worker_num} --image ${image} --command "${command}"
```

Thank you!

**Comment by guillaumekln** [comment]
Do you have more logs to know what happens every 50 steps? Is an evaluation running?

**Comment by Morizeyao** [comment]
No I didn't set the script to do any evaluation, and there are no logs when stuck. 

By the way, I just ran some experiments and see that when I set sample_buffer_size to 0, there seems to be no stuck problem.

I will continue doing experiments and update my progress here. :)

**Comment by Morizeyao** [comment]
For now my solution is disabling shuffle (set sample_buffer_size to 0) and manually repeat and shuffle the training data several times before sending them to training. The disk space is not a problem for now.

**Comment by guillaumekln** [comment]
So the slowdown may happen when workers are refilling the shuffle buffer. Does the training also take a long time to start? Do all workers have sufficient system memory?

**Comment by Morizeyao** [comment]
Yes it does take long time to start, and each worker (with 4 P100 GPUs) has 40GB memory.

Some good news here: after setting sample_buffer_size to 0, the training seems to run smoothly on distributed setting! It seems that the problem comes from shuffling process. Maybe because lack of memory?

**Comment by guillaumekln** [comment]
40GB of system memory should be fine.

I'm not familiar with HDFS though. I see that TensorFlow has some recommendations regarding reading remote datasets: https://www.tensorflow.org/guide/data_performance#parallelizing_data_extraction. Let's see if we can integrate them at some point.

**Comment by Morizeyao** [comment]
Yea I'll do some research too. Closing this issue for now as it is mostly solved.

If I have more findings I'll share them here. :)
